### DISCUSSION TITLE
Multiple Module Sections in TSL - What is the proper coded workflow

### DISCUSSION THREAD

**Discussion by TaviTruman** [microsoft/GraphEngine#356] [General]
@shaobin Dr. Shao, what is the correct coded workflow on the GE Client side when multiple "module" declarations exist in the TSL. Thinking logically, each module should be implemented via a separate .NET assembly, while on the GE Client side, the client could load one server module segment at a time. Is this correct?

**Comment by shaobin** [comment]
@TaviTruman Could you provide a simple scenario with a simplified TSL to show what you want to do?

**Reply by TaviTruman** [reply]
TSL Module Declarations:

I have one library where these modules are defined:

Here is my VS Project Structure:

![image](https://user-images.githubusercontent.com/5692812/182212960-3424f78a-b3a4-4aa3-8ec4-08c312fa9245.png)

module ContactGraphDataModelServerModule
{
  protocol PushContactDataModelAsync;
  protocol SaveContactDataModelAsync;
  protocol RemoveContactDataModelAsync;
  protocol UpdateContactDataModelAsync;
  protocol QueryContactDataModelAsync;  
}

module ServiceObjectsGraphDataServerModule
{
  protocol SaveLeadValidationData;
  protocol FetchLeadValidationData;
  protocol RemoveLeadValidationData;
  protocol UpdateLeadValidationData;
  protocol PushLeadValidationData;
}

// Generate Trinity Modular Server Protocol Interfaces
module UserAuthenticationGraphDataServerModule
{
  protocol SetUserAuthenticationDataAsync;
  protocol PostSetUserAuthRequestComplete;
  protocol FetchUserAuthenticationDataAsync;
  protocol QueryUserAuthInfoAsync;
  protocol UserAuthInfoQueryAnswerResponseAsync;
}

As you know the TSL Compiler generates the "Protocols.cs" file where the RPC Functions are stored.  These server module implementations are bound to the host library

In my server-side code I have these statements:

I have been experimenting with this code as you can see some lines of code are commented out.

![image](https://user-images.githubusercontent.com/5692812/182216486-bb2a34f6-1ad2-447b-85f1-7e0814304d78.png)

Here is the code I have on the Client-side:

![image](https://user-images.githubusercontent.com/5692812/182217210-ea48a968-6259-4a80-92fb-6bc6ba975021.png)

This coded workflow does not seem to work as the RPC Message IDs get crossed and the calls are indexed into the wrong RPC code segments. As a workaround, I create separate assemblies for each module but that defeats the original design of the "module" concept.

So, I think I'm not using the API correctly.

**Reply by TaviTruman** [reply]
@shaobin Dr. Shao do you need a full set of working TSL code?
